### PR TITLE
emscripten: Upgrade emsdk to 3.1.68

### PR DESCRIPTION
**Pull request by kleisauke**
In line with commit https://github.com/rust-lang/rust/commit/2c38ecfc9077000db7de14bf501fa5294eeecedd.

Resolves: #3306.
Supersedes: #3569.
Cherry-picks a couple of changes from: #3307.

### CODE REVIEW

**Comment by hoodmane** [comment]
Thanks @kleisauke it would be nice to see time_t finally get fixed...

**Comment by juntyr** [comment]
Could/should this PR keep the emcc version check to give a nice error at build time if a now-no-longer supported Emscripten is used?

**Comment by hoodmane** [comment]
I wouldn't want rust to generate an error when building with Emscripten 3.1.58 since that is the version that Pyodide is using with Python 3.12 and it would prevent us from upgrading. Though I suppose that it would just require us to use a patched `rust/libc` and we are already doing that, it would be nice to avoid additional patches.

**Comment by hoodmane** [comment]
But having a version check would be fine as long as there is an override.

**Comment by juntyr** [comment]
I was thinking about the previous check for 3.1.42 when the ABI change happened, which previous libc versions still supported (now there would be an ABI break in the opposite direction).

I definitely agree that other than that pretty much any emcc version has to be allowed since there is no current indicator which versions are ABI compatible and which aren’t

**Comment by tgross35** [comment]
How backwards compatible is emsdk? I think putting this on the 0.2 stable branch should be okay, but would like to know if there is any expected breakage here.

**Comment by kleisauke** [comment]
Updated now that PR https://github.com/rust-lang/rust/pull/131533 was landed. This is ready for review.

> How backwards compatible is emsdk?

AFAIK, there were no ABI-breaking changes after Emscripten 3.1.42 (released at 2023/06/23), which was addressed in libc by commit 63b0d673eaf2a177ff208c5c50999738bdd1bf0d (but only when building with `-Zbuild-std`).

**Comment by kleisauke** [comment]
> Could/should this PR keep the emcc version check to give a nice error at build time if a now-no-longer supported Emscripten is used?

I could address this within a follow-up PR, but perhaps its severity should be a warning instead? Simple programs that don't use filesystem operations will likely still work with Emscripten versions prior to 3.1.42. On the other hand, Emscripten 3.1.42 is already more than a year old.

**Comment by hoodmane** [comment]
> AFAIK, there were no ABI-breaking changes after Emscripten 3.1.42

@kleisauke it would be really great if we could make some sort of tool to do ABI compatibility checks. For a static test, we could compare the contents of `emscripten/src/struct_info_generated.json`. Perhaps we could use `emscripten/tools/maint/gen_struct_info.py` to generate info about the layout of more standard library structs than just the ones that are in the default struct_info. For a dynamic test, maybe take the unistd test suite and build the tests as side modules with one emcc version then load them and call into them with a second emcc version. 

This probably wouldn't be perfect, but it would be a lot better than the current method where we don't really know anything.

File: src/unix/linux_like/emscripten/mod.rs
Comment (tgross35) [inline src/unix/linux_like/emscripten/mod.rs:869]: It is awesome to have the emscripten PRs linked, but could you put them all into the commit message and delete the commented-out code? Since these are things we want to keep in history but it isn't relevant to the code anymore.

**Comment by tgross35** [comment]
I don't have any major concerns here but I'll defer to @hoodmane's review (I'll merge once you give it the checkmark)

File: src/unix/linux_like/emscripten/mod.rs
Comment (kleisauke) [reply src/unix/linux_like/emscripten/mod.rs]: Good idea! Split into commit 476bf225658143f5689fa2986e6fb491c8081eaf and 99f8c8751f94d6149e55c6d238f310303694ca4c.

File: libc-test/build.rs
Comment (hoodmane) [inline libc-test/build.rs:2733]: Most of these went away in https://github.com/emscripten-core/emscripten/pull/17704/ but I don't see `aio.h` or `sys/sysctl.h`. Were they removed in a separate PR?

File: src/unix/linux_like/emscripten/align.rs
Comment (hoodmane) [inline src/unix/linux_like/emscripten/align.rs:43]: (Irrelevant remark.) The stack still has to be aligned to 16 bytes when you call a function even though maxalign is only 8.  Discovering this wasted a lot of my time.

File: libc-test/build.rs
Comment (hoodmane) [inline libc-test/build.rs:2924]: What does this FIXME mean? Did that PR break this test? Why?

**Review by hoodmane — APPROVED** [review]
Thanks @kleisauke generally looks good to me.

File: libc-test/build.rs
Comment (kleisauke) [reply libc-test/build.rs:2733]: This was removed via PR https://github.com/emscripten-core/emscripten/pull/18058. Hence, I've split this removal into commit 476bf225658143f5689fa2986e6fb491c8081eaf.

Comment (kleisauke) [reply libc-test/build.rs:2733]: Oh, indeed `sys/sysctl.h` has been removed via https://github.com/emscripten-core/emscripten/pull/18863. Let me split that change as well.

File: libc-test/build.rs
Comment (kleisauke) [reply libc-test/build.rs:2924]: That PR did break the libc test, see e.g.:
```
RUNNING ALL TESTS
bad SIG_IGN value at byte 0: rust: 1 (0x1) != c 254 (0xfe)
bad SIG_IGN value at byte 1: rust: 0 (0x0) != c 255 (0xff)
bad SIG_IGN value at byte 2: rust: 0 (0x0) != c 255 (0xff)
bad SIG_IGN value at byte 3: rust: 0 (0x0) != c 255 (0xff)
thread 'main' panicked at /checkout/target/wasm32-unknown-emscripten/debug/build/libc-test-890b1e2d4768b06a/out/main.rs:12:21:
some tests failed
```

However, this `FIXME` can probably not fixed according to 
https://github.com/emscripten-core/emscripten/pull/14883/files#r892912370.

I'll change this to:
```suggestion
            // `SIG_IGN` has been changed to -2 since 1 is a valid function address
            // https://github.com/emscripten-core/emscripten/pull/14883
            "SIG_IGN" => true,
```

Note that other implementations have a similar issue, see e.g:
https://github.com/rust-lang/libc/blob/e4c7accc68b1a376eb1f538d7b991802793afe96/libc-test/build.rs#L1198

File: build.rs
Comment (kleisauke) [reply build.rs:235]: This change was made based on the assumption that Rust requires Emscripten 3.1.42 or later. However, upon review, it appears that the compatibility note in https://github.com/rust-lang/rust/issues/131689 indicates that versions prior 3.1.42 are still supported.

Let me drop commit a35f2120b95413a787edff15411d789416c5bbbf for now.

File: libc-test/build.rs
Comment (kleisauke) [reply libc-test/build.rs:2733]: Fixed with commit fbeecf3e812b05a4357d0e2ef663766f8cbe2b2b.

File: libc-test/build.rs
Comment (hoodmane) [reply libc-test/build.rs]: Maybe there should be a constant somewhere that indicates what the value of SIG_IGN is on the platform... But that would presumably be dealt with best in a followup.

File: src/unix/linux_like/emscripten/align.rs
Comment (tgross35) [inline src/unix/linux_like/emscripten/align.rs:41]: This can just go in the commit message

File: src/unix/linux_like/emscripten/mod.rs
Comment (tgross35) [inline src/unix/linux_like/emscripten/mod.rs:751]: These look like they are octal in the musl source, could they stay that way?

**Review by tgross35 — CHANGES_REQUESTED** [review]
Just a couple small nits before merge. Thanks @hoodmane for taking a look.

**Comment by tgross35** [comment]
One last thing: could you add the following section to README.md after "Rust version support" so we have this noted somewhere? I wanted to document everything we interface with at some point, might as well start here.

```markdown
## Interface version support

`libc` targets being able to interface with the following minimum library
versions:

- Emscripten 3.1.42, released 2023-06-23
```

File: src/unix/linux_like/emscripten/align.rs
Comment (kleisauke) [reply src/unix/linux_like/emscripten/align.rs]: Split into commit 52fa133223a5a338494f1abab42e2a47a464e0d6 and moved the `time_t` description to the commit message as well.
https://github.com/rust-lang/libc/compare/b6a957f2ec9f24a843d04ec4313f3a4634944072..a37fa9d267ddb3c7cbc9504ac730a7ed6fc11f59

File: src/unix/linux_like/emscripten/mod.rs
Comment (kleisauke) [reply src/unix/linux_like/emscripten/mod.rs]: I reverted this part, I can't remember why I did this in the first place.

**Comment by kleisauke** [comment]
@rustbot ready

> One last thing: could you add the following section to README.md after "Rust version support" so we have this noted somewhere? [...]

I think this should be done as a follow-up given that this PR does not raise the minimum Emscripten version (at least when building with `-Zbuild-std`), see:
https://github.com/rust-lang/libc/pull/3962#discussion_r1832494361

**Pull request merged**
